_posts/function-contracts.md

```diff
@@ -0,0 +1,405 @@
+
+Lets first look at the harness that tests `TokenBucket::new`. It is straightforward, setting up a series of non-deterministic inputs, calling the function under verification and then asserting a validity condition.
+
+```rust
+#[kani::proof]
+fn verify_token_bucket_new() {
+    let size = kani::any_where(|s| *s != 0);
+    let one_time_burst = kani::any();
+    let complete_refill_time_ms = kani::any_where(|t|
+        *t != 0 && *t < u64::MAX / NANOSEC_IN_ONE_MILLISEC
+    );
+
+    let bucket = TokenBucket::new(size, one_time_burst, complete_refill_time_ms).unwrap()
+    assert!(bucket.is_valid());
+}
+```
+
+Since the most important call here is to `TokenBucket::new` let’s take a closer look at its implementation.
+
+```rust
+pub fn new(size: u64, one_time_burst: u64, complete_refill_time_ms: u64) -> Option<Self> {
+    if size == 0 || complete_refill_time_ms == 0 {
+        return None;
+    }
+    let complete_refill_time_ns =
+        complete_refill_time_ms.checked_mul(NANOSEC_IN_ONE_MILLISEC)?;
+    let common_factor = gcd(size, complete_refill_time_ns);
+    let processed_capacity: u64 = size / common_factor;
+    let processed_refill_time: u64 = complete_refill_time_ns / common_factor;
+
+    Some(TokenBucket {
+        size,
+        one_time_burst,
+        initial_one_time_burst: one_time_burst,
+        refill_time: complete_refill_time_ms,
+        budget: size,
+        last_update: Instant::now(),
+        processed_capacity,
+        processed_refill_time,
+    })
+}
+```
+
+Most of what happens here is just a series of divisions, with the exception of this call: `gcd(size, complete_refill_time_ns)` so let us look at that function next:
+
+```rust
+fn gcd(mut max: u64, mut min: u64) -> u64 {
+    if min > max {
+        std::mem::swap(&mut max, &mut min);
+    }
+
+    let rest = max % min;
+    if rest == 0 { min } else { gcd(min, rest) }
+}
+
+```
+
+This is by far the costliest part of the code of `TokenBucket::new`. The rest of the function was a fixed set of divisions, but here we have a division and a recursive call back to `gcd`. It is not immediately obvious how busy this computation is but in the [worst case](https://en.wikipedia.org/wiki/Euclidean_algorithm#Worst-case) the number of steps (recursions) approaches 1.5 times the number of bits needed to represent the input numbers. Meaning that for two large 64 bit numbers it can take almost 96 iterations for a single call to `gcd`. Recall that our smallest input to `TokenBucket::new` is a non-deterministic value in the range $0< x < 18446744073709$ (up to 45 non-zero bits). If a harness that uses `gcd` does not heavily constrain the input, Kani would have to unroll the recursion at least 68 times and then execute it symbolically. An expensive operation.
```

Comment:
```suggestion
This is by far the costliest part of the code of `TokenBucket::new`. The rest of the function was a fixed set of divisions, but here we have a division and a recursive call back to `gcd`. It is not immediately obvious how busy this computation is but in the [worst case](https://en.wikipedia.org/wiki/Euclidean_algorithm#Worst-case) the number of steps (recursions) approaches 1.5 times the number of bits needed to represent the input numbers. Meaning that for two large 64 bit numbers it can take almost 96 iterations for a single call to `gcd`. Recall that our smallest input to `TokenBucket::new` is a non-deterministic value in the range $0< x < 18446744073709$ (up to 45 non-zero bits). If a harness that uses `gcd` does not heavily constrain the input, Kani would have to unroll the recursion at least 68 times and then execute it symbolically. An expensive operation.
```

Reply:
@roypat If you have time, do you think you could help us by reviewing this post? Thank you!

Reply:
Hi @celinval, sure I can have a look! I might only get some time next week though :(

Reply:
I have no idea either now 😅 I guess we can leave it.